File: console/src/api/configs.ts
import type { Bootstrap, TenantRow } from "@/api/core";
import type { CodingAgentKind } from "@/domain/codingAgent";
import type {
  ApplicationStatus,
  AuthPropagationPreviewResponse,
  AuthPropagationReport,
  ConfigCatalogEntry,
  ConfigFileResponse as GeneratedConfigFileResponse,
  ConfigListResponse as GeneratedConfigListResponse,
  ConfigAuthResponse as GeneratedConfigAuthResponse,
  LinkedConfigFileResponse as GeneratedLinkedConfigFileResponse,
  CustomProviderState as GeneratedCustomProviderState,
  VisualConfigOptionState as GeneratedVisualConfigOptionState,
  DiagnoseConfigResponse,
  LastApplication,
  PropagationOutcome,
} from "@/api/generated/wire";
import { listTenantsRequest } from "@/api/tenants";
import type { ControlApi } from "@/api/transport";
import { tenantBody, tenantQuery } from "@/api/tenantSelection";
import type { TenantSelection } from "@/domain/tenant";

export type { ApplicationStatus, ConfigCatalogEntry, LastApplication, PropagationOutcome };

export type ConfigListData = GeneratedConfigListResponse;

export type ConfigLinkedFileData = GeneratedLinkedConfigFileResponse;

export type ConfigCustomProvider = GeneratedCustomProviderState;

export type ConfigAuthData = Omit<GeneratedConfigAuthResponse, "mode"> & {
  mode: "chatgpt" | "api-key";
};

export type ConfigVisualOption = Omit<GeneratedVisualConfigOptionState, "value" | "value_kind"> & {
  value_kind: "string" | "bool";
  value?: string | boolean;
  proxy_routed: boolean;
};

export type ConfigFileData = Omit<
  GeneratedConfigFileResponse,
  "visual_options" | "custom_provider" | "visual_error" | "warnings" | "auth" | "linked_file"
> & {
  visual_options?: ConfigVisualOption[];
  custom_provider?: ConfigCustomProvider;
  visual_error?: string;
  warnings?: string[];
  auth?: ConfigAuthData;
  linked_file?: ConfigLinkedFileData;
};

export interface ConfigFileTarget {
  tenant: TenantSelection;
  agent: CodingAgentKind;
  current: boolean;
  config: string | null;
  file: string;
}

export type ConfigFileDiagnostics = DiagnoseConfigResponse;

export interface ConfigFileInput {
  revision: string;
  contentBase64: string;
  visualOptions?: Array<{ path: string; included: boolean; value?: string | boolean }>;
  customProvider?: { included: boolean; name: string; base_url: string; proxy_routed: boolean };
  visualAuth?: { included: boolean; value: string };
}

export type PropagationPreview = AuthPropagationPreviewResponse;

export type PropagationReport = AuthPropagationReport;

export interface ConfigApi {
  bootstrap: Bootstrap;
  listTenants(signal?: AbortSignal): Promise<TenantRow[]>;
  listConfigs(
    tenant: TenantSelection,
    agent: CodingAgentKind,
    signal?: AbortSignal,
  ): Promise<ConfigListData>;
  revealConfigFile(target: ConfigFileTarget): Promise<ConfigFileData>;
  diagnoseConfigFile(
    target: ConfigFileTarget,
    contentBase64: string,
  ): Promise<ConfigFileDiagnostics>;
  saveConfigFile(target: ConfigFileTarget, input: ConfigFileInput): Promise<ConfigFileData>;
  createConfig(tenant: TenantSelection, agent: CodingAgentKind, name: string): Promise<void>;
  applyConfig(tenant: TenantSelection, agent: CodingAgentKind, name: string): Promise<void>;
  deleteConfigs(tenant: TenantSelection, agent: CodingAgentKind, names: string[]): Promise<void>;
  previewCredentialPropagation(): Promise<PropagationPreview>;
  executeCredentialPropagation(planId: string): Promise<PropagationReport>;
}

function configTargetBody(target: ConfigFileTarget) {
  return {
    ...tenantBody(target.tenant),
    agent: target.agent,
    current: target.current,
    config: target.config,
    file: target.file,
  };
}

function visualOption(value: GeneratedVisualConfigOptionState): ConfigVisualOption {
  if (value.value_kind !== "string" && value.value_kind !== "bool") {
    throw new Error(`Unsupported Visual Config value kind: ${value.value_kind}`);
  }
  if (
    value.value !== null &&
    value.value !== undefined &&
    typeof value.value !== "string" &&
    typeof value.value !== "boolean"
  ) {
    throw new Error(`Unsupported Visual Config value for ${value.path}`);
  }
  const { value: rawValue, ...rest } = value;
  return {
    ...rest,
    value_kind: value.value_kind,
    proxy_routed: false,
    ...(rawValue === null || rawValue === undefined ? {} : { value: rawValue }),
  };
}

function authData(value: GeneratedConfigAuthResponse): ConfigAuthData {
  if (value.mode !== "chatgpt" && value.mode !== "api-key") {
    throw new Error(`Unsupported Config auth mode: ${value.mode}`);
  }
  return { ...value, mode: value.mode };
}

function configFileData(value: GeneratedConfigFileResponse): ConfigFileData {
  return {
    file: value.file,
    exists: value.exists,
    revision: value.revision,
    content_base64: value.content_base64,
    visual_options: value.visual_options?.map(visualOption) ?? undefined,
    custom_provider: value.custom_provider ?? undefined,
    visual_error: value.visual_error ?? undefined,
    warnings: value.warnings,
    auth: value.auth ? authData(value.auth) : undefined,
    linked_file: value.linked_file ?? undefined,
  };
}

export function configsApi(client: ControlApi): ConfigApi {
  return {
    bootstrap: client.bootstrap,
    listTenants: listTenantsRequest(client),
    listConfigs: (tenant, agent, signal) => {
      const query = tenantQuery(tenant);
      query.set("agent", agent);
      return client.get<GeneratedConfigListResponse>(`/_aibox/api/configs?${query}`, signal);
    },
    revealConfigFile: (target) =>
      client
        .post<GeneratedConfigFileResponse>("/_aibox/api/configs/reveal", configTargetBody(target))
        .then(configFileData),
    diagnoseConfigFile: (target, contentBase64) =>
      client.post<DiagnoseConfigResponse>("/_aibox/api/configs/diagnose", {
        ...configTargetBody(target),
        content_base64: contentBase64,
      }),
    saveConfigFile: (target, input) =>
      client
        .post<GeneratedConfigFileResponse>("/_aibox/api/configs/save", {
          ...configTargetBody(target),
          revision: input.revision,
          content_base64: input.contentBase64,
          ...(input.visualOptions ? { visual_options: input.visualOptions } : {}),
          ...(input.customProvider ? { custom_provider: input.customProvider } : {}),
          ...(input.visualAuth ? { visual_auth: input.visualAuth } : {}),
        })
        .then(configFileData),
    createConfig: async (tenant, agent, config) => {
      await client.post("/_aibox/api/configs/create", { ...tenantBody(tenant), agent, config });
    },
    applyConfig: async (tenant, agent, config) => {
      await client.post("/_aibox/api/configs/apply", { ...tenantBody(tenant), agent, config });
    },
    deleteConfigs: async (tenant, agent, configs) => {
      await client.post("/_aibox/api/configs/delete", {
        ...tenantBody(tenant),
        agent,
        configs,
        all: false,
        confirmation: configs.length === 1 ? configs[0] : "",
      });
    },
    previewCredentialPropagation: () =>
      client.post<PropagationPreview>("/_aibox/api/configs/propagate-auth/preview"),
    executeCredentialPropagation: (planId) =>
      client.post<PropagationReport>("/_aibox/api/configs/propagate-auth/execute", {
        plan_id: planId,
      }),
  };
}

File: src/service/control/configs.rs
//! Config Control API handlers and wire types.

use super::{
    AgentTenantQuery, ControlResult, default_agent, default_tenant_selection, json_response,
};
use crate::agent::AgentKind;
use crate::application_error::{ApplicationErrorKind, application_error};
use crate::config::{self, CustomProviderInput, VisualConfigOptionInput};
use crate::service::coordination::{ConfigCoordinator, ConfigFileView, DeleteConfigsCommand};
use crate::service::state::ServiceState;
use crate::tenant::TenantSelection;
use axum::Json;
use axum::extract::{Query, State};
use axum::http::StatusCode;
use base64::Engine as _;
use serde::{Deserialize, Serialize};
use serde_json::Value;

#[derive(Serialize)]
#[cfg_attr(test, derive(ts_rs::TS))]
pub(crate) struct ConfigListResponse {
    configs: Vec<config::ConfigCatalogEntry>,
    files: &'static [&'static str],
    application: config::ApplicationStatus,
    credential_propagation_available: bool,
}

/// Decode a base64 wire field, keeping the original Control API wording.
fn decode_base64(value: &str) -> Result<Vec<u8>, anyhow::Error> {
    base64::engine::general_purpose::STANDARD
        .decode(value.as_bytes())
        .map_err(|error| {
            application_error(
                ApplicationErrorKind::InvalidInput,
                format!("invalid base64: {error}"),
            )
        })
}

pub(super) async fn list_configs(
    State(state): State<ServiceState>,
    Query(query): Query<AgentTenantQuery>,
) -> ControlResult {
    let selection = TenantSelection::parse(&query.tenant)?;
    let catalog = ConfigCoordinator::new(state)
        .list(selection, query.agent)
        .await?;
    Ok(json_response(
        StatusCode::OK,
        &ConfigListResponse {
            configs: catalog.configs,
            files: catalog.files,
            application: catalog.application,
            credential_propagation_available: catalog.credential_propagation_available,
        },
    ))
}

#[derive(Serialize)]
#[cfg_attr(test, derive(ts_rs::TS))]
pub(crate) struct AuthPropagationPreviewResponse {
    plan_id: String,
    preview: config::AuthPropagationPreview,
}

pub(super) async fn preview_auth_propagation(
    State(state): State<ServiceState>,
    Json(_request): Json<Value>,
) -> ControlResult {
    let preview = ConfigCoordinator::new(state)
        .preview_auth_propagation()
        .await?;
    Ok(json_response(
        StatusCode::OK,
        &AuthPropagationPreviewResponse {
            plan_id: preview.plan_id,
            preview: preview.preview,
        },
    ))
}

#[derive(Deserialize)]
#[cfg_attr(test, derive(ts_rs::TS))]
pub(crate) struct ExecuteAuthPropagationRequest {
    plan_id: String,
}

pub(super) async fn execute_auth_propagation(
    State(state): State<ServiceState>,
    Json(request): Json<ExecuteAuthPropagationRequest>,
) -> ControlResult {
    let report = ConfigCoordinator::new(state)
        .execute_auth_propagation(request.plan_id)
        .await?;
    Ok(json_response(StatusCode::OK, &report))
}

#[derive(Deserialize)]
#[cfg_attr(test, derive(ts_rs::TS))]
#[serde(deny_unknown_fields)]
pub(crate) struct ConfigMutationBase {
    #[serde(default = "default_tenant_selection")]
    tenant: String,
    #[serde(default = "default_agent")]
    agent: AgentKind,
    config: String,
}

pub(super) async fn create_config(
    State(state): State<ServiceState>,
    Json(request): Json<ConfigMutationBase>,
) -> ControlResult {
    let selection = TenantSelection::parse(&request.tenant)?;
    let name = config::NamedConfigName::parse(&request.config)?;
    let created = ConfigCoordinator::new(state)
        .create(selection, request.agent, name)
        .await?;
    Ok(json_response(
        StatusCode::OK,
        &CreatedConfigResponse { created },
    ))
}

#[derive(Deserialize)]
#[cfg_attr(test, derive(ts_rs::TS))]
#[serde(deny_unknown_fields)]
pub(crate) struct ConfigFileRequest {
    #[serde(default = "default_tenant_selection")]
    tenant: String,
    #[serde(default = "default_agent")]
    agent: AgentKind,
    #[serde(default)]
    current: bool,
    config: Option<String>,
    file: String,
}

#[derive(Serialize)]
#[cfg_attr(test, derive(ts_rs::TS))]
pub(crate) struct ConfigFileResponse {
    file: String,
    exists: bool,
    revision: String,
    content_base64: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    visual_options: Option<Vec<config::VisualConfigOptionState>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    custom_provider: Option<crate::config::CustomProviderState>,
    #[serde(skip_serializing_if = "Option::is_none")]
    visual_error: Option<String>,
    #[serde(skip_serializing_if = "Vec::is_empty", default)]
    warnings: Vec<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    auth: Option<ConfigAuthResponse>,
    #[serde(skip_serializing_if = "Option::is_none")]
    linked_file: Option<LinkedConfigFileResponse>,
}

#[derive(Serialize)]
#[cfg_attr(test, derive(ts_rs::TS))]
pub(crate) struct LinkedConfigFileResponse {
    file: String,
    exists: bool,
    revision: String,
    content_base64: String,
}

#[derive(Serialize)]
#[cfg_attr(test, derive(ts_rs::TS))]
pub(crate) struct ConfigAuthResponse {
    mode: &'static str,
    api_key: Option<String>,
    extra_fields: bool,
    warnings: Vec<String>,
}

pub(super) async fn reveal_config_file(
    State(state): State<ServiceState>,
    Json(request): Json<ConfigFileRequest>,
) -> ControlResult {
    let selection = TenantSelection::parse(&request.tenant)?;
    let target = config::ConfigTarget::from_wire(request.config.as_deref(), request.current)?;
    let file = config::ConfigFile::parse(request.agent, &request.file)?;
    let view = ConfigCoordinator::new(state)
        .reveal(selection, request.agent, target, file)
        .await?;
    Ok(json_response(StatusCode::OK, &config_file_response(view)))
}

#[derive(Deserialize)]
#[cfg_attr(test, derive(ts_rs::TS))]
#[serde(deny_unknown_fields)]
pub(crate) struct SaveConfigFileRequest {
    #[serde(default = "default_tenant_selection")]
    tenant: String,
    #[serde(default = "default_agent")]
    agent: AgentKind,
    #[serde(default)]
    current: bool,
    config: Option<String>,
    file: String,
    revision: String,
    content_base64: String,
    #[serde(default)]
    visual_options: Option<Vec<VisualConfigOptionInput>>,
    #[serde(default)]
    custom_provider: Option<CustomProviderInput>,
    #[serde(default)]
    visual_auth: Option<crate::config::VisualAuthInput>,
}

pub(super) async fn save_config_file(
    State(state): State<ServiceState>,
    Json(request): Json<SaveConfigFileRequest>,
) -> ControlResult {
    let selection = TenantSelection::parse(&request.tenant)?;
    let content = decode_base64(&request.content_base64)?;
    let target = config::ConfigTarget::from_wire(request.config.as_deref(), request.current)?;
    let file = config::ConfigFile::parse(request.agent, &request.file)?;
    let edit = config::ConfigEdit::from_wire(
        content,
        request.custom_provider,
        request.visual_options,
        request.visual_auth,
    )?;
    let view = ConfigCoordinator::new(state)
        .save(
            selection,
            request.agent,
            target,
            file,
            request.revision,
            edit,
        )
        .await?;
    Ok(json_response(StatusCode::OK, &config_file_response(view)))
}

#[derive(Deserialize)]
#[cfg_attr(test, derive(ts_rs::TS))]
#[serde(deny_unknown_fields)]
pub(crate) struct DiagnoseConfigRequest {
    #[serde(default = "default_tenant_selection")]
    tenant: String,
    #[serde(default = "default_agent")]
    agent: AgentKind,
    #[serde(default)]
    current: bool,
    config: Option<String>,
    file: String,
    content_base64: String,
}

/// One Config syntax error as the Console reads it.
///
/// This mirrors [`config::ConfigDiagnostic`] rather than reusing it because the
/// wire shape carries a `severity` the domain has no notion of. Where a wire
/// shape matches its domain type exactly, pass the domain type straight through
/// instead — `ProtocolSummary` does — and keep a second
/// name only for a genuine difference like this one. A field added to the
/// domain type has to be added here too before it reaches the Console.
#[derive(Serialize)]
#[cfg_attr(test, derive(ts_rs::TS))]
pub(crate) struct ConfigDiagnostic {
    severity: &'static str,
    message: String,
    line: usize,
    column: usize,
}

#[derive(Serialize)]
#[cfg_attr(test, derive(ts_rs::TS))]
pub(crate) struct DiagnoseConfigResponse {
    diagnostics: Vec<ConfigDiagnostic>,
}

pub(super) async fn diagnose_config_file(
    State(state): State<ServiceState>,
    Json(request): Json<DiagnoseConfigRequest>,
) -> ControlResult {
    let selection = TenantSelection::parse(&request.tenant)?;
    let content = decode_base64(&request.content_base64)?;
    let target = config::ConfigTarget::from_wire(request.config.as_deref(), request.current)?;
    let file = config::ConfigFile::parse(request.agent, &request.file)?;
    let diagnostics = ConfigCoordinator::new(state)
        .diagnose(selection, request.agent, target, file, content)
        .await?;
    Ok(json_response(
        StatusCode::OK,
        &DiagnoseConfigResponse {
            diagnostics: diagnostics
                .into_iter()
                .map(|diagnostic| ConfigDiagnostic {
                    severity: "error",
                    message: diagnostic.message,
                    line: diagnostic.line,
                    column: diagnostic.column,
                })
                .collect(),
        },
    ))
}

pub(super) async fn apply_config(
    State(state): State<ServiceState>,
    Json(request): Json<ConfigMutationBase>,
) -> ControlResult {
    let selection = TenantSelection::parse(&request.tenant)?;
    let name = config::NamedConfigName::parse(&request.config)?;
    let application = ConfigCoordinator::new(state)
        .apply(selection, request.agent, name)
        .await?;
    Ok(json_response(StatusCode::OK, &application))
}

#[derive(Deserialize)]
#[cfg_attr(test, derive(ts_rs::TS))]
#[serde(deny_unknown_fields)]
pub(crate) struct DeleteConfigsRequest {
    #[serde(default = "default_tenant_selection")]
    tenant: String,
    #[serde(default = "default_agent")]
    agent: AgentKind,
    #[serde(default)]
    configs: Vec<String>,
    #[serde(default)]
    all: bool,
    confirmation: String,
}

pub(super) async fn delete_configs(
    State(state): State<ServiceState>,
    Json(request): Json<DeleteConfigsRequest>,
) -> ControlResult {
    let selection = TenantSelection::parse(&request.tenant)?;
    let command = DeleteConfigsCommand {
        selection,
        agent: request.agent,
        configs: request.configs,
        all: request.all,
        confirmation: request.confirmation,
    };
    let deleted = ConfigCoordinator::new(state).delete(command).await?;
    Ok(json_response(
        StatusCode::OK,
        &DeletedConfigsResponse {
            deleted: deleted.configs,
            all: deleted.all,
        },
    ))
}

#[derive(Serialize)]
#[cfg_attr(test, derive(ts_rs::TS))]
pub(crate) struct CreatedConfigResponse {
    created: String,
}

#[derive(Serialize)]
#[cfg_attr(test, derive(ts_rs::TS))]
pub(crate) struct DeletedConfigsResponse {
    deleted: Vec<String>,
    all: bool,
}

fn config_file_response(view: ConfigFileView) -> ConfigFileResponse {
    let visual_error = view.visual.error;
    let (visual_options, custom_provider) = view.visual.state.map_or((None, None), |state| {
        (Some(state.options), state.custom_provider)
    });
    let auth = view.auth.map(|auth| ConfigAuthResponse {
        mode: auth.mode,
        api_key: auth.api_key,
        extra_fields: auth.extra_fields,
        warnings: auth.warnings,
    });
    ConfigFileResponse {
        file: view.snapshot.file,
        exists: view.snapshot.exists,
        revision: view.snapshot.revision,
        content_base64: base64::engine::general_purpose::STANDARD.encode(view.snapshot.content),
        visual_options,
        custom_provider,
        visual_error,
        warnings: view.warnings,
        auth,
        linked_file: view.linked.map(linked_config_file_response),
    }
}

fn linked_config_file_response(snapshot: config::ConfigFileSnapshot) -> LinkedConfigFileResponse {
    LinkedConfigFileResponse {
        file: snapshot.file,
        exists: snapshot.exists,
        revision: snapshot.revision,
        content_base64: base64::engine::general_purpose::STANDARD.encode(snapshot.content),
    }
}
